Sources/TFMongoSwift/Dataset.swift

```diff
@@ -0,0 +1,222 @@
+            return opts.batchSize!
+        }
+    }
+
+    fileprivate func execute(on collection: MongoCollection<Document>) throws -> MongoCursor<Document> {
+        switch self {
+        case let .find(filter, opts):
+            if let filter = filter {
+                return try collection.find(filter, options: opts)
+            }
+            return try collection.find(options: opts)
+        case let .aggregate(pipeline, opts):
+            return try collection.aggregate(pipeline, options: opts)
+        }
+    }
+}
+
+/// Dataset backed by a lazily evaluated cursor. Each batch makes a round trip to the server. Each iteration of the
+/// iterator yields a `TensorGroup` converted from the first generic type to the second.
+public struct MongoDataset<C: Codable, T: TensorGroup & InitializableFromSequence>: Sequence where T.SequenceType == C {
+    public typealias Iterator = MongoDatasetIterator<C, T>
+    public typealias Element = T
+
+    private let client: MongoClient
+    private let collection: MongoCollection<Document>
+    private var query: Query
+
+    private var batchSize: Int32 {
+        get {
+            return self.query.batchSize
+        }
+        set(newSize) {
+            switch self.query {
+            case let .find(filter, opts):
+                let newOpts = FindOptions(batchSize: newSize, limit: opts.limit, projection: opts.projection)
+                self.query = .find(filter: filter, opts: newOpts)
+            case let .aggregate(pipeline, _):
+                let newOpts = AggregateOptions(batchSize: newSize)
+                self.query = .aggregate(pipeline: pipeline, opts: newOpts)
+            }
+        }
+    }
+
+    /// Initialize a `Dataset` from the result of the provided MongoDB aggregation pipeline.
+    /// - SeeAlso: https://docs.mongodb.com/manual/aggregation/
+    public init(uri: String? = nil, db: String, collection: String, pipeline: [Document]) throws {
+        if let uri = uri {
+            self.client = try MongoClient(uri)
+        } else {
+            self.client = try MongoClient()
+        }
+        self.collection = client.db(db).collection(collection)
+        self.query = .aggregate(pipeline: pipeline, opts: AggregateOptions(batchSize: 1))
+    }
+
+    /// Initialize a `Dataset` from the given MongoDB collection, optionally providing a filter, projection, or limit
+    /// on the number of results.
+    public init(uri: String? = nil,
+                db: String,
+                collection: String,
+                filter: Document? = nil,
+                projection: Document? = nil,
+                limit: Int64? = nil) throws {
+        if let uri = uri {
+            self.client = try MongoClient(uri)
+        } else {
+            self.client = try MongoClient()
+        }
+
+        self.collection = self.client.db(db).collection(collection)
+        let opts = FindOptions(batchSize: 1, limit: limit, projection: projection)
```

Comment:
do we necessarily want/need the `MongoCursor` batch size and the TensorFlow batch size to match up? 
I guess I don't have a good sense of how people use batching in TensorFlow, but if the batch size is usually a very small number/1 this may be forcing them into a lot of network round trips. 
it seems possible that people would want to separately configure these and let the server retrieve some amount > 1 at a time.

Reply:
I think you're right. For now, I've changed the TensorFlow batching to be separate and not provided a way to configure the cursor batching. I don't think users will need to configure how much the cursor actually retrieves since it'll be a manageable amount with the default setting.